packages/flutter/lib/src/widgets/lazy_block.dart

```diff
@@ -0,0 +1,449 @@
+  double _startOffsetUpperLimit = 0.0;
+
+  @override
+  void visitChildren(ElementVisitor visitor) {
+    for (Element child in _children)
+      visitor(child);
+  }
+
+  @override
+  void mount(Element parent, dynamic newSlot) {
+    super.mount(parent, newSlot);
+    renderObject
+      ..callback = _layout
+      ..mainAxis = widget.direction;
+    // Children will get built during layout.
+    // Paint offset will get updated during layout.
+  }
+
+  @override
+  void update(LazyBlockViewport newWidget) {
+    LazyBlockViewport oldWidget = widget;
+    super.update(newWidget);
+    renderObject.mainAxis = widget.direction;
+    if (newWidget.delegate.runtimeType != oldWidget.delegate.runtimeType
+        || newWidget.delegate.shouldRebuild(oldWidget.delegate)) {
+      IndexedBuilder builder = widget.delegate.buildItem;
+      List<Widget> widgets = new List<Widget>(_children.length);
+      for (int i = 0; i < widgets.length; ++i)
+        widgets[i] = builder(this, _firstChildLogicalIndex + i);
+      _children = updateChildren(_children, widgets);
+    }
+    if (widget.startOffset >= _startOffsetLowerLimit && widget.startOffset < _startOffsetUpperLimit)
+      _updatePaintOffset();
+    else
+      renderObject.markNeedsLayout();
+  }
+
+  @override
+  void unmount() {
+    renderObject.callback = null;
+    super.unmount();
+  }
+
+  void _layout(BoxConstraints constraints) {
+    owner.lockState(() {
+      _doLayout(constraints);
+    }, building: true, context: 'during $runtimeType layout');
+  }
+
+  void _doLayout(BoxConstraints constraints) {
+    final IndexedBuilder builder = widget.delegate.buildItem;
+    final double blockExtent = _getMainAxisExtent(renderObject.size);
+    final double startLogicalOffset = widget.startOffset;
+    final double endLogicalOffset = startLogicalOffset + blockExtent;
+    final _RenderLazyBlock block = renderObject;
+    final BoxConstraints innerConstraints = _getInnerConstraints(constraints);
+
+    // A high watermark for which children have been through layout this pass.
+    int firstLogicalIndexNeedingLayout = _firstChildLogicalIndex;
+
+    // The index of the current child we're examining. The index is the same one
+    // used for the builder (as opposed to the physical index in the _children
+    // list).
+    int currentLogicalIndex = _firstChildLogicalIndex;
+
+    // The offset of the current child we're examining from the start of the
+    // entire block (in the direction of the main axis). As we compute layout
+    // information, we use dead reckoning to keep track of where all the
+    // children are based on this quantity.
+    double currentLogicalOffset = _firstChildLogicalOffset;
+
+    // First, we check if we need to inflate any children before the start of
+    // the viewport. Because we're dead reckoning from the current viewport, we
+    // inflate the children in reverse tree order.
+
+    if (currentLogicalIndex > 0 && currentLogicalOffset > startLogicalOffset) {
```

Comment:
This implies that this is only a half-infinite scrolling list. Is that intentional?